activerecord/lib/active_record/connection_adapters/abstract/query_log_tags.rb

```diff
@@ -0,0 +1,227 @@
+          comments.each do |comment|
+            if comment.present? && !sql.include?(comment)
+              sql = if prepend_comment
+                "#{comment} #{sql}"
+              else
+                "#{sql} #{comment}"
+              end
+            end
+          end
+          sql
+        end
+      end
+
+      module ExecutionMethods
+        def execute(sql, *args, **kwargs)
+          super(self.class.add_query_log_tags_to_sql(sql), *args, **kwargs)
+        end
+
+        def exec_query(sql, *args, **kwargs)
+          super(self.class.add_query_log_tags_to_sql(sql), *args, **kwargs)
+        end
+      end
+
+      # Maintains a user-defined context for all queries and constructs an SQL comment by
+      # calling methods listed in +components+
+      #
+      # Additional information can be added to the context to be referenced by methods
+      # defined in framework or application initializers.
+      #
+      # To add new comment components, add a hash to the components array containing the keys and values you
+      # want to add to the comment. Dynamic content can be created by setting a proc or lambda value in a hash,
+      # and can reference any value stored in the +context+ object.
+      #
+      # Example:
+      #
+      #    components = [
+      #      :application,
+      #      { custom_component: -> { context[:controller].controller_name } }
+      #    ]
+      #    ActiveRecord::ConnectionAdapters::QueryLogTags::QueryLogTagsContext.components = components
+      #    ActiveRecord::ConnectionAdapters::QueryLogTags::QueryLogTagsContext.comment
+      #    # /*application:MyApp,custom_component:ProjectsController*/
+      #
+      # Alternatively, for more complex comment construction, define class methods on +QueryLogTagsContext+ in
+      # your application.
+      #
+      #    module ActiveRecord::ConnectionAdapters::QueryLogTags::QueryLogTagsContext
+      #      class << self
+      #        def custom_component
+      #          "custom value"
+      #        end
+      #      end
+      #    end
+      #    ActiveRecord::ConnectionAdapters::QueryLogTags::QueryLogTagsContext.components = [:custom_component]
+      #    ActiveRecord::ConnectionAdapters::QueryLogTags::QueryLogTagsContext.comment
+      #    # /*custom_component:custom value*/
+      #
+      # Default components available for use:
+      #
+      # * +application+
+      # * +pid+
+      # * +socket+
+      # * +db_host+
+      # * +database+
+      # * +line+ (reported via BacktraceCleaner)
+      #
+      # _When included in Rails, ActiveController and ActiveJob components are also defined._
+      #
+      # * +controller+
+      # * +action+
+      # * +job+
+      #
+      # If required due to log truncation, comments can be prepended to the query instead:
+      #
+      #    ActiveRecord::ConnectionAdapters::QueryLogTags.prepend_comment = true
```

Comment:
is it worth also documenting the `cache_query_log_tags` option here? i think it's the only thing not documented already. it's off by default, and on reading the code it's not immediately obvious to me when or why you'd want to use it.

Reply:
Great point - will do so, and expose it as a core config option too.

Reply:
Resolving as this has been documented 🙇‍♂️